### ISSUE TITLE
Option to 'spread out' color scheme

### ISSUE THREAD

**Issue by peterszerzo** [chartist-js/chartist#256]
I am currently using chartist with a SASS color scheme comprised of 15 colors, ranging from dark purple to light blue. If I only have 3 data series, the corresponding svg elements get the subclasses .ct-series-a, .ct-series-b and .ct-series-c, that is, they will be assigned colors 1, 2 and 3. If there was a way to have Chartist spread things out and use colors 1, 8 and 15 instead (or in more general terms, interpolate based on the number of series for each individual chart on the page), the visual separation of the chart would be much better.

Gion, do you think you may want to have this as an option in the library { spreadColorScheme: true }? I would be happy to come up with an implementation and issue a pull request.

All in all, wonderful project, really happy with it!

**Comment by gionkunz** [comment]
Hi @pickled-plugins 

The thing on Sass and CSS is that you can assign whichever colours you want. If you know you'll have 3 colours then you can choose them in your styling.

If you're more talking about a programmatic way of setting the colours, I'd not do this via the class names. This would make the class names non-semantic and represent a colour instead of a series.

However, what I can think of would be great to have, for all those dynamically created charts that should use colours based on spread, data value (intensity, heat map etc), would be a colour plugin that allows you to do this and that will set the colour via style attribute.

What do you think?

**Comment by peterszerzo** [comment]
Hi Gion,

Setting only three colors that spread our color scheme makes sense if we had a single chart in our webpage. Right now, we are planning on using Chartist for several charts, some with 3 series, some with 10, some with 15 - and we are looking for a 'one-size-fits-all' solution. Having a finely graded color scheme is something we'd like to keep - this way, for charts with many series, we'll have visual cues such as 'the bluer the color, the later the year', or something of the sort.

Powering color assignments with a color plug-in would be great. The only other solution I can think of is moving the responsibility of color assignments out of the ".ct-series" class to honor the semantics and having a new ".ct-series-color-#{some-number}" class, where JS would determine the some-number variable for a given series by doing the interpolation (1 - 8 - 15 vs. 1 - 2 - 3 from my original post). For my current project, I've implemented this latter workflow with a d3 heatmap - it leaves colors within the stylesheets (which visual designers - like ours - may prefer) and we found it simpler to implement, though most likely more limited than a plug-in.

Thoughts?

**Comment by gionkunz** [comment]
Hi Peter

I'm kind of a hardliner when it comes to clean separation of concerns, and if I find a colour in some JavaScript code I usually write a letter of complaint to the W3C consortium. However, for some things, like the case you mentioned, I think it makes absolute sense to do the colour configuration in JavaScript. Even if you have classes which reflect stages of a colouring process, you'd not want to rely on discrete values right? Ideally you'd want to have continuous values that fully represent the state of the data with a individual color, shape, whatever. For these kind of things I guess there is no other way  than doing this dynamically in JavaScript.

Chartist is designed to do things like this with its custom events that allows you to intercept the process of Chartist drawing the charts.

There were already several cases where I agreed this makes absolute sense.
https://github.com/gionkunz/chartist-js/issues/79#issuecomment-65908224
Or
https://github.com/gionkunz/chartist-js/issues/124#issuecomment-69713687

I'd really appreciate a universal colour plugin for Chartist. Some ideas:
- Taking data value as input to determine either colour intensity, hue, saturation, brightness or all of them
- Taking some additional meta data on value objects for determining the colour
- Spreading the colour based on the index and total amount of series

More ideas?

**Comment by peterszerzo** [comment]
I personally would not mind the discrete color values (we did some visual experiments with our d3 heat maps and are happy with the results), and the folks who prefer not to touch JavaScript will probably agree. This method could be documented as an optional, inaccurate, 'quick-and-dirty' way for those purists out there. Implementation will be much quicker as well. And with that, I'll stop pushing :).

Plugin-wise - there is also good research on how to interpolate colors to achieve good visual separation, and I have a sense Chroma.js have it figured out pretty well (https://vis4.net/blog/posts/avoid-equidistant-hsv-colors/). Optional dependency? I'll prototype something this weekend if I get to it.

Many thanks for your input!

**Comment by peterszerzo** [comment]
One more idea:
1. Add the following two variables in the SASS sheet:
$ct-series-colorscheme-start: black;
$ct-series-colorscheme-end: white;
2. Create two divs with classes that apply these colors on them. Hide them from view.
3. Read off these colors in JavaScript. Now we have them inside the script while still being able to customize them in the stylesheets.
4. Let the not-yet-born color plug-in do its thing.

**Comment by gionkunz** [comment]
@pickled-plugins , although I really like the idea of a colour plugin, I really don't have the time to start something right now. If you start something in a plugin and make it open, I'll sure review or maybe even contribute.

**Comment by peterszerzo** [comment]
Great, I'll be in touch.

**Comment by gionkunz** [comment]
@pickled-plugins Did you go any further with chartist and your colour plugin?

**Comment by peterszerzo** [comment]
Hi Gion,

Given time constraints for a project, we ended up applying custom interpolated colors using jQuery after the charts are already created by Chartist, something along the lines of 

```
$chartLines.each(function() {$(this).css('stroke', /* interpolation logic referenced here */ }); 
```

The workflow was part of an open-source library https://github.com/pickled-plugins/chartist-html that my intern developed. The library takes a specified html syntax that contains chart data, parses this syntax, builds Chartist charts for each qualifying html snippet, applies coloring, binds and unbinds tooltips (for single-page Backbone app cleanup), applies label formatting specified in the HTML (percent, currency, date...), and solves certain layout issues we couldn't resolve with chartist (such as really really long labels on the y-axis that should take up no more than 3 lines).

Long story short, the library is very project-specific, and requires substantial code quality improvement and further documentation. We will get to this in the coming weeks, at which point we'd like to ask you to review.

**Comment by peterszerzo** [comment]
Hi Gion,

I had a color plug-in idea that I'd like to run by you (happy to start implementing based on your feedback - have time for it too!).  It came to me after I looked at the coloring example on your website - http://gionkunz.github.io/chartist-js/getting-started.html#colour-up-your-charts.

Here is the way it works: inside chart options, one could simply supply a color function that takes as an argument the same context that is passed onto the 'draw' event in the example.

```
var options = {
    width: 20, height: 30, /* and so on */
    color: function colorFunction(context) { return 'rgb(10, 20, 30);' }
};
```

Then, in your library, when the chart element is constructed (before 'draw' is triggered), we can use this function to work out the color and apply it with all the other attributes and styling (width, height, position). This would eliminate the need for listening to the 'draw' event and then manipulating the dom for the second time. It would make the syntax more declarative.

At this point, we could use an outside color library to handle the inner workings of `colorFunction`. Are you interested in having a lightweight version of such a library inside Chartist?

The context has pretty much all the information one would possibly need for any complex coloring feature, except the series index (if I have a 3 line charts on the same graph, the indeces 0, 1, 2 does not show up, which is not a problem for pie charts). If we included it, things would be pretty simple from that point on.

What do you think?

**Comment by peterszerzo** [comment]
@gionkunz could you give me your thoughts on my last message above? I am sorry I've been so laggy about this - I'll be out of work in the month of December (moving back to Europe in a couple of weeks), and I'd like to help out with this.

**Comment by gionkunz** [comment]
@pickled-plugins , no worries I didn't had time to work on Chartist lately either. I'd prefer a separate plugin for this. Plugins can also easily be configured within Chartist:

``` javascript
var chart = new Chartist.Line('.ct-chart', {
  labels: [1, 2, 3, 4, 5, 6, 7],
  series: [
    [1, 5, 3, 4, 6, 2, 3],
    [2, 4, 2, 5, 4, 3, 6]
  ]
}, {
  plugins: [
    ctColourPlugin({
      baseColor: ...,
      spreadFactor: ...,
      ...
    })
  ]
});
```

The series index was included in the events with 485ae62

**Comment by daN4cat** [comment]
In #451 I suggested a way to colouring pie slices using a random #rgb generation.

I don't think a plugin is really required given the flexibility offered by the on draw hook.

**Comment by oleksiy-nesterov** [comment]
@pickled-plugins, here is small plugin example to get color from list, or directly from series (series:[{color:'#000'...}, {...}])

```js
(function(window, document, Chartist){
    'use strict';
    var defaultOptions = {
        colors:[]
    };
    Chartist.plugins = Chartist.plugins || {};
    Chartist.plugins.colors = function(options){
        options = Chartist.extend({}, defaultOptions, options);
        return function colors(chart){
            chart.on('draw', function(data){
                if(data.type === 'bar' || data.type === 'slice'){
                    var color = options.colors[data.seriesIndex] || data.series.color;
                    color && data.element.attr({style:[data.type === 'slice' ? 'fill' : 'stroke', ':', color, ';'].join('')});
                };
            });
        };
    };
}(window, document, Chartist));
```
